### PR TITLE
Add RealmEncryptionNotSupportedException

### PR DESCRIPTION
**Pull request by beeender**
- [x] More comments
- [x] Core header file change

### CODE REVIEW

File: realm/src/main/java/io/realm/exceptions/RealmEncryptionNotSupportedException.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package io.realm.exceptions;
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/exceptions/RealmEncryptionNotSupportedException.java]: Missing License

**Comment by beeender** [comment]
@realm/java Review, thanks. Core change is coming.

File: realm/src/main/java/io/realm/Realm.java [hunk 2]
```diff
@@ -198,6 +198,8 @@ protected void checkIfValid() {
      * @param configuration Configuration used to open the Realm.
      * @param autoRefresh {@code true} if Realm should auto-refresh. {@code false} otherwise.
      * @throws IllegalArgumentException if trying to open an encrypted Realm with the wrong key.
+     * @throws io.realm.exceptions.RealmEncryptionNotSupportedException if the device doesn't
+     * support Realm encryption for some reasons.
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/Realm.java]: Remove "for some reasons", it is just boilerplate :)

File: realm/src/main/java/io/realm/Realm.java [hunk 3]
```diff
@@ -375,6 +377,8 @@ public static Realm getDefaultInstance() {
      *
      * @throws RealmMigrationNeededException If no migration has been provided by the configuration and the
      * model classes or version has has changed so a migration is required.
+     * @throws io.realm.exceptions.RealmEncryptionNotSupportedException if the device doesn't
+     * support Realm encryption for some reasons.
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/Realm.java]: Same

**Comment by cmelchior** [comment]
I think we should also add a line to `RealmConfiguration.Builder.encryptionKey()` about this as that is really the most obvious place people interact with encryption. Perhaps something along the lines of:

```
...

Note that some devices do not support the encryption used by Realm. These devices will instead throw a {@link RealmEncryptionNotSupportedException } when the Realm is opened. See {@link RealmEncryptionNotSupportedException } for further details.
```

File: realm/src/main/java/io/realm/exceptions/RealmEncryptionNotSupportedException.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+package io.realm.exceptions;
+
+/**
+ * On some devices (HTC One X for example), for some reason, the system doesn't pass the right
+ * parameter (siginfo_t.si_addr) to the segfault signal handler which our encryption mechanism
+ * relies on. Realm will try to detect if this problem exists on the device when an encrypted
+ * Realm is being created/opened. A RealmEncryptionNotSupportedException will be thrown if this
+ * problem exists which causes encryption cannot be done on this device.
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/exceptions/RealmEncryptionNotSupportedException.java]: which means that encryption cannot be used on this device.

File: changelog.txt [hunk 5]
```diff
@@ -3,9 +3,10 @@
  * Fixed a bug which might cause failure when loading the native library.
  * Fixed a bug which might trigger a timeout in Context.finalize().
  * Fixed a bug which might cause RealmObject.isValid() to throw an exception if the object is deleted.
- * Updated Realm core to version 0.89.7
+ * Updated Realm core to version 0.89.8
    - fixed a potential stack overflow issue which might cause a crash when encryption was used.
    - embedded crypto functions from OpenSSL into Realm dynamic lib to avoid random issues on some devices.
+   - throw RealmEncryptionNotSupportedException if the device doesn't support Realm encryption.
```
Comment (bmunkholm) [inline changelog.txt]: Add details here, like: 
At least one device type (HTX One X) contain system bugs that prevents Realms encryption to function properly. This is now detected, and an exception is thrown when trying to open/create an encrypted Realm file. It's up to the application to catch this and decide if it's ok to proceed without encryption instead.

File: realm/src/main/java/io/realm/Realm.java [hunk 6]
```diff
@@ -375,6 +376,7 @@ public static Realm getDefaultInstance() {
      *
      * @throws RealmMigrationNeededException If no migration has been provided by the configuration and the
      * model classes or version has has changed so a migration is required.
+     * @throws io.realm.exceptions.RealmEncryptionNotSupportedException if the device doesn't support Realm encryption.
```
Comment (bmunkholm) [inline realm/src/main/java/io/realm/Realm.java]: For the other exceptions we don't include "io.realm.exceptions."

File: realm/src/main/java/io/realm/RealmConfiguration.java [hunk 7]
```diff
@@ -281,6 +281,10 @@ public Builder name(String filename) {
 
         /**
          * Sets the 64 bit key used to encrypt and decrypt the Realm file.
+         * <p>
+         * Note that some devices do not support the encryption used by Realm. These devices will instead throw a
```
Comment (bmunkholm) [inline realm/src/main/java/io/realm/RealmConfiguration.java]: "Note that some devices..." -> "Note that a few older devices.." 
Which ones have we identified? I think we should list them (somewhere - maybe not here) so people can decide if they want to support them.

File: realm/src/main/java/io/realm/exceptions/RealmEncryptionNotSupportedException.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.exceptions;
+
+/**
+ * On some devices (HTC One X for example), for some reason, the system doesn't pass the right
```
Comment (bmunkholm) [inline realm/src/main/java/io/realm/exceptions/RealmEncryptionNotSupportedException.java]: Have we identified others than HTC One X? If so we just list that and mention we have only heard it for that device.

Comment (beeender) [inline realm/src/main/java/io/realm/exceptions/RealmEncryptionNotSupportedException.java]: HTC One X+ maybe. I guess Alcatel crash reported by pinterest might be the same root cause. But I don't have evidence to prove that.

**Comment by bmunkholm** [comment]
otherwise :+1:

File: changelog.txt [hunk 9]
```diff
@@ -3,9 +3,10 @@
  * Fixed a bug which might cause failure when loading the native library.
  * Fixed a bug which might trigger a timeout in Context.finalize().
  * Fixed a bug which might cause RealmObject.isValid() to throw an exception if the object is deleted.
- * Updated Realm core to version 0.89.7
-   - fixed a potential stack overflow issue which might cause a crash when encryption was used.
-   - embedded crypto functions from OpenSSL into Realm dynamic lib to avoid random issues on some devices.
+ * Updated Realm core to version 0.89.8
+   - Fixed a potential stack overflow issue which might cause a crash when encryption was used.
+   - Embedded crypto functions from OpenSSL into Realm dynamic lib to avoid random issues on some devices.
```
Comment (kneth) [inline changelog.txt]: Do we really wish to say something about implementation details? I think we should remove "from OpenSSL".

Comment (beeender) [inline changelog.txt]: @bmunkholm Your opinion? For people who cares can easily find out we are using functions from OpenSSL i think...

Comment (beeender) [inline changelog.txt]: And it is in the git history already :(

File: realm/src/main/java/io/realm/exceptions/RealmEncryptionNotSupportedException.java (hunk 8 above)
Comment (kneth) [inline realm/src/main/java/io/realm/exceptions/RealmEncryptionNotSupportedException.java]: Isn't the list of unsupported devices a good candidate for a FAQ?

Comment (beeender) [inline realm/src/main/java/io/realm/exceptions/RealmEncryptionNotSupportedException.java]: In that case, we also need to put all the details like AOSP version, ROM version, all those information. For example, I believe HTC One X with another ROM (CM maybe) won't have this problem. This incompatibility is software issue which caused by the OEM but not the phone itself.

File: changelog.txt (hunk 9 above)
Comment (kneth) [inline changelog.txt]: I'm not against letting people know what we are using (in git log). But I find we should leave out implementation details in the change log.

Comment (bmunkholm) [inline changelog.txt]: Yeah, don't think "from OpenSSL" adds any useful information/value to the sentence. So fine to remove.

File: realm/src/main/java/io/realm/exceptions/RealmEncryptionNotSupportedException.java (hunk 8 above)
Comment (bmunkholm) [inline realm/src/main/java/io/realm/exceptions/RealmEncryptionNotSupportedException.java]: My main point is that this seems very rare and therefore users needs to have an idea about how rare this is and that they should expect encryption to work in almost all cases. Do we have some stat's from the users involved? How many devices in % of their total were affected?

**Comment by beeender** [comment]
retest this please

File: changelog.txt [hunk 10]
```diff
@@ -3,10 +3,11 @@
...
  * Fixed a bug which might trigger a timeout in Context.finalize().
  * Fixed a bug which might cause RealmObject.isValid() to throw an exception if the object is deleted.
- * Updated Realm core to version 0.89.7
-   - fixed a potential stack overflow issue which might cause a crash when encryption was used.
-   - embedded crypto functions from OpenSSL into Realm dynamic lib to avoid random issues on some devices.
  * Fixed an issue preventing the call of listeners on refresh().
+ * Updated Realm core to version 0.89.8
+   - Fixed a potential stack overflow issue which might cause a crash when encryption was used.
+   - Embedded crypto functions into Realm dynamic lib to avoid random issues on some devices.
+   - Throw RealmEncryptionNotSupportedException if the device doesn't support Realm encryption. At least one device type (HTX One X) contains system bugs that prevents Realm's encryption to function properly. This is now detected, and an exception is thrown when trying to open/create an encrypted Realm file. It's up to the application to catch this and decide if it's OK to proceed without encryption instead.
```
Comment (cmelchior) [inline changelog.txt]: from functioning properly

**Comment by cmelchior** [comment]
Kinda massive blob to put in a changelog, but having it being really explicit is properly good. Apart from minor detail in changelog :+1:

**Comment by bmunkholm** [comment]
Please add the exception here: https://github.com/realm/realm-java/blob/4ac64c87aaca494083bd76a3ec27c8a3fdee7ab6/realm/src/main/java/io/realm/internal/Util.java

And likely a comment in the jni code where excetions are defined, that we have to remember to add a test here.

**Comment by beeender** [comment]
@bmunkholm Done.

**Comment by bmunkholm** [comment]
:+1:

**Comment by cotfas** [comment]
Hello

I just wanted to ask

If the phone model doesn`t support encryption (ex htconex+)

When I call 

 realm.writeEncryptedCopyTo(newFile, newKey);

Will RealmEncryptionNotSupportedException will be thrown?

Thanks

**Comment by cotfas** [comment]
​@beeender 

I`ve tested the new jar

On getInstance the RealmEncryptionNotSupportedException  is thrown on my htconex

But when calling 
realm.writeEncryptedCopyTo(newFile, newKey);

No exception has been thrown

But when I`m going to files

The new encrypted file has 0kb


Here is a snapshot 

                        Log.e("REALM" ,"writeEncryptedCopyTo BEFORE realmNotEncrypted.writeEncryptedCopyTo(newFile, newKey) ");
                        realmNotEncrypted.writeEncryptedCopyTo(newFile, newKey);

                        Log.e("REALM", "writeEncryptedCopyTo AFTER realmNotEncrypted.writeEncryptedCopyTo(newFile, newKey) ");
                        realmNotEncrypted.close();

                        Log.e("REALM", "writeEncryptedCopyTo AFTER    realmNotEncrypted.close()");

                        Realm.deleteRealm(configNotEncrypted);

                        Log.e("REALM", "writeEncryptedCopyTo STOP");


Here is the log


09-03 18:46:11.591    2312-2312/com.cashdivider.app E/REALM﹕ writeEncryptedCopyTo BEFORE realmNotEncrypted.writeEncryptedCopyTo(newFile, newKey)
09-03 18:46:11.591    2312-2312/com.cashdivider.app A/libc﹕ Fatal signal 11 (SIGSEGV) at 0x00000000 (code=2), thread 2312 (cashdivider.app) - sent from sys



I`m not sure if I can detect if the htconex doesn`t support encryption because first I have to migrate the unencrypted to encrypted state and after that calling realm.getInstance to throw the RealmEncryptionNotSupportedException

**Pull request merged**